internal/saucecloud/cloud.go

```diff
@@ -934,3 +967,128 @@ func (r *CloudRunner) getHistory(launchOrder config.LaunchOrder) (insights.JobHi
+	// Skip reporting if job is not completed
+	if !job.Done(res.job.Status) || res.skipped || res.job.ID == "" {
+		return
+	}
+
+	assets, err := r.JobService.GetJobAssetFileNames(context.Background(), res.job.ID, res.job.IsRDC)
+	if err != nil {
+		log.Warn().Err(err).Str("action", "loadAssets").Msg(msg.InsightsReportError)
+		return
+	}
+
+	var testRuns []insights.TestRun
+
+	if arrayContains(assets, saucereport.SauceReportFileName) {
+		report, err := r.loadSauceTestReport(res.job.ID, res.job.IsRDC)
+		if err != nil {
+			log.Warn().Err(err).Str("action", "parsingJSON").Msg(msg.InsightsReportError)
+			return
+		}
+		testRuns = insights.FromSauceReport(report)
+	} else if arrayContains(assets, junit.JunitFileName) {
+		report, err := r.loadJUnitReport(res.job.ID, res.job.IsRDC)
+		if err != nil {
+			log.Warn().Err(err).Str("action", "parsingXML").Msg(msg.InsightsReportError)
+			return
+		}
+		testRuns = insights.FromJUnit(report)
+	}
+	enrichInsightTestRun(testRuns, res.job.ID, res.name, res.details, res.job.IsRDC)
+	if len(testRuns) > 0 {
+		if err := r.InsightsService.PostTestRun(context.Background(), testRuns); err != nil {
+			log.Warn().Err(err).Str("action", "posting").Msg(msg.InsightsReportError)
+		}
+	}
+}
+
+func (r *CloudRunner) loadSauceTestReport(jobID string, isRDC bool) (saucereport.SauceReport, error) {
+	fileContent, err := r.JobService.GetJobAssetFileContent(context.Background(), jobID, saucereport.SauceReportFileName, isRDC)
+	if err != nil {
+		log.Warn().Err(err).Str("action", "loading-json-report").Msg(msg.InsightsReportError)
+		return saucereport.SauceReport{}, err
+	}
+	return saucereport.Parse(fileContent)
+}
+
+func (r *CloudRunner) loadJUnitReport(jobID string, isRDC bool) (junit.TestSuites, error) {
+	fileContent, err := r.JobService.GetJobAssetFileContent(context.Background(), jobID, junit.JunitFileName, isRDC)
+	if err != nil {
+		log.Warn().Err(err).Str("action", "loading-xml-report").Msg(msg.InsightsReportError)
+		return junit.TestSuites{}, err
+	}
+	return junit.Parse(fileContent)
+}
+
+func arrayContains(list []string, want string) bool {
+	for _, item := range list {
+		if item == want {
+			return true
+		}
+	}
+	return false
+}
+
+func enrichInsightTestRun(runs []insights.TestRun, jobID string, jobName string, details insightDetails, isRDC bool) {
+	var ciData ci.CI
+	provider := ci.GetProvider()
+	ciData = ci.GetCI(provider)
+
+	for idx := range runs {
+		runs[idx].Browser = details.Browser
+		runs[idx].BuildName = details.BuildName
+		runs[idx].Device = resolveDevice(details.DeviceName, details.DeviceID)
+		runs[idx].Framework = details.Framework
+		runs[idx].OS = resolveOS(details.Platform, details.Framework)
+		runs[idx].Platform = resolvePlatform(isRDC)
+		runs[idx].SauceJob = &insights.Job{
+			ID:   jobID,
+			Name: jobName,
+		}
+		runs[idx].Tags = details.Tags
+		runs[idx].Type = resolveType(details.Framework)
+
+		if ci.HasProvider(provider) {
```

Comment:
Not sure how much value this function provides, since you can do
```suggestion
		if provider != None {
```

Reply:
Actually, that was derived from here: https://github.com/saucelabs/saucectl/blob/main/internal/ci/ci.go#L258

But will correct, I agree that's easier to read !